- 정상적인 프로그램의 흐름을 방해하는 사건

- 주로 프로그램 오류, 버그 때문에 발생한다.

- 예외 발생 시

  - 예외 상황을 복구해서 정상적인 흐름으로 전환 가능한가?

    - 재시도

      - 외부와 연결된 경우 외부의 문제로 예외가 발생할 수 있다. 이 경우 시간을 좀 기다렸다가 다시 시도하면 정상 수행되는 경우가 있다.

    - 대안

      - 두 개 이상의 서로 다른 대안을 준비해서 하나가 안되면 다른 대안으로 시도한다.

  - 버그인가?

    - 예외가 발생한 코드의 버그인가?

    - 클라이언트의 버그인가?

      이 경우는 클라이언트에게 제대로 요청을 보내라고 말을 해서 해결할 수 있다.

  - 제어할 수 없는 예외 상황인가?

- 예외를 잘못 다루는 코드

  - 예외를 무시하고 아무런 조치도 하지 않는 코드

    ```java
    try {
        ...
    }
    catch(~~~) {
    }
    ```

  - 무의미하고 무책임한 throws

    - 계속 밖으로 예외를 던지기만 하면 나중에 어디서 예외가 발생했는지 찾기 어렵다.

      <center>
      <img
          src="https://github.com/user-attachments/assets/7395ae59-773a-431b-a5bc-4ad07ccaa604"
          width="100%"
      />
      </center>

- 예외의 종류

  - Error

    - 시스템에 비정상적인 상황이 발생

    - OutOfMemoryError

    - ThreadDeath

  - 체크 예외 (Exception)

    - 우리가 신경써야하는 예외다.

    - catch, throws를 통해 예외를 다뤄야 함.

    - 복구할 수 없다면 RuntimeException이나 적절한 추상화 레벨의 예외로 전환해서 던져야 한다.

  - 예외의 추상화와 전환

    - 사용 기술에 따라 서로 다른 예외가 던져진다. 이렇게되면 특정 기술을 사용하다가 다른 기술로 변경되면 예외를 다루는 코드에서 수정이 발생하게 된다. 그래서 적절한 예외 추상화와 예외 번역이 필요하다.
